File: python/examples/multimodal.py
"""Live multimodal example: attach an image to a run, and let a tool hand one back.

Runs the shared 8x8 fixture (`examples/media/fixture.png`, quadrants red/green/blue/
white) through both provider styles against OpenRouter, and closes tasks 12.1-12.3 of
the `add-multimodal-content` change for this port.

    OPENROUTER_API_KEY=... python examples/multimodal.py

**Arrival is proven by the prompt-token delta, never by the model's answer.** A model
asked to name colours will happily name four colours it never received — which is
exactly how the silent-drop bug this release fixes stayed hidden. It cannot fake
prompt tokens.

Two things are measured per style:

1. attachment — the identical request without and with `image(fixture)`; the prompt
   tokens must jump.
2. §8A relocation — a tool returning an image on `ToolResult.parts`, called by the
   model. The run must complete AND cost materially more prompt tokens than the same
   run whose tool returns text only. On `openai` the image is relocated into a
   synthetic user message (a `tool` message cannot carry one); on `anthropic` it rides
   inside `tool_result`. Both are exercised here, live.

Cheap by construction: tiny models, an 82-byte image, `max_tokens` capped at 40.
The API key is read from the environment and is never printed, logged or written.

Known upstream defect (see CHANGELOG.md): OpenRouter accepts an image bound for an
Anthropic model with HTTP 200 and drops it en route — a ~+4 token delta instead of
hundreds. Observed on both of its endpoints (`/chat/completions` with an openai-shaped
`image_url`, and the Anthropic-compatible `/v1/messages` with a native `source{}`
block), and reproducible with plain curl carrying no toolnexus code at all. It is a
routing defect above this library, so it is reported as `image=dropped-upstream`,
never as a failure of toolnexus. To exercise an Anthropic model with a working image
path, point `base_url` at `https://api.anthropic.com` with an `ANTHROPIC_API_KEY`.
"""
from __future__ import annotations

import asyncio
import os

from toolnexus import ToolResult, create_client, create_toolkit, define_tool, image

_HERE = os.path.dirname(os.path.abspath(__file__))
_EXAMPLES = os.path.normpath(os.path.join(_HERE, "..", "..", "examples"))
FIXTURE = os.path.join(_EXAMPLES, "media", "fixture.png")

ASK = (
    "Name the four quadrant colours of this image, clockwise from top-left. "
    "Answer with four words only."
)
TOOL_ASK = (
    "Call the screenshot tool, then name the four quadrant colours of the image it "
    "returns, clockwise from top-left. Answer with four words only."
)
COLOURS = ("red", "green", "blue", "white")

# Even this 82-byte image costs hundreds of prompt tokens wherever it actually
# arrives (8 500 on gpt-4o-mini, 263 on gemini-2.5-flash-lite — a tile budget, not a
# byte count). Anything under this is the image having been dropped en route; a
# double-digit difference is turn-to-turn noise, not an image.
MIN_IMAGE_TOKENS = 200

STYLES = (
    ("openai", "openai/gpt-4o-mini"),
    ("anthropic", "anthropic/claude-haiku-4.5"),
)


def _colours_named(text: str) -> int:
    low = text.lower()
    return sum(1 for c in COLOURS if c in low)


def _screenshot_tool(*, with_image: bool):
    """A tool returning an 8x8 screenshot — with or without the image part.

    The parts-less twin is the control: the prompt-token difference between the two
    runs is the image, and nothing else.
    """
    part = image(FIXTURE)

    def screenshot() -> ToolResult:
        """Capture the current screen and return it as a PNG."""
        return ToolResult(
            output="screenshot captured, 8x8 png",
            is_error=False,
            parts=[{"type": part.type, "mimeType": part.mimeType, "data": part.data}]
            if with_image
            else None,
        )

    return define_tool(screenshot, name="screenshot")


def _client(style: str, model: str, key: str):
    return create_client(
        base_url="https://openrouter.ai/api/v1",
        style=style,
        model=model,
        api_key=key,
        request_params={"max_tokens": 40},
    )


async def _run_style(style: str, model: str, key: str) -> str:
    agent = _client(style, model, key)

    # --- 1. attachment: the same request, without and with the image ---------
    # builtins off: their ~2 000-token schema would drown the signal we measure
    bare = await create_toolkit(builtins=False)
    text_only = await agent.run(ASK, bare)
    with_image = await agent.run([ASK, image(FIXTURE)], bare)
    ptok_text = text_only.usage["prompt_tokens"]
    ptok_image = with_image.usage["prompt_tokens"]
    delta = ptok_image - ptok_text
    arrived = delta >= MIN_IMAGE_TOKENS
    colours = _colours_named(with_image.text)
    print(f"\n[{style}] {model}")
    print(f"  text-only ptok={ptok_text}  with-image ptok={ptok_image}  delta=+{delta}")
    print(f"  answer: {with_image.text.strip()[:120]!r}  ({colours}/4 colours named)")
    if not arrived:
        print("  ^ image did NOT arrive: too few prompt tokens. Upstream drop, not a")
        print("    toolnexus failure — the block is emitted per SPEC §8A either way.")
    await bare.close()

    # --- 2. §8A relocation: a tool that returns an image ---------------------
    tk_img = await create_toolkit(builtins=False, extra_tools=[_screenshot_tool(with_image=True)])
    tk_txt = await create_toolkit(builtins=False, extra_tools=[_screenshot_tool(with_image=False)])
    res_img = await agent.run(TOOL_ASK, tk_img)
    res_txt = await agent.run(TOOL_ASK, tk_txt)
    await tk_img.close()
    await tk_txt.close()

    called = [c["name"] for c in res_img.tool_calls]
    reloc_delta = res_img.usage["prompt_tokens"] - res_txt.usage["prompt_tokens"]
    # Two independent facts: the loop completed with the part in it (ours), and the
    # image actually reached the model (upstream's to drop).
    reloc = "ok" if bool(res_img.text) and "screenshot" in called else "failed"
    reloc_image = "ok" if reloc_delta >= MIN_IMAGE_TOKENS else "dropped-upstream"
    print(
        f"  tool calls: {called}  turns={res_img.turns}/{res_txt.turns}  "
        f"tool-result ptok delta=+{reloc_delta}  -> loop={reloc} image={reloc_image}"
    )
    print(f"  answer: {res_img.text.strip()[:120]!r}")

    return (
        f"RESULT python style={style} model={model} ptok_text={ptok_text} "
        f"ptok_image={ptok_image} delta=+{delta} image={'ok' if arrived else 'dropped-upstream'} "
        f"colours={colours}/4 relocation={reloc} reloc_image={reloc_image} "
        f"reloc_delta=+{reloc_delta}"
    )


async def main() -> None:
    key = os.environ.get("OPENROUTER_API_KEY")
    if not key:
        print("(no OPENROUTER_API_KEY — skipping the live multimodal run)")
        return

    lines = [await _run_style(style, model, key) for style, model in STYLES]
    print()
    for line in lines:
        print(line)


if __name__ == "__main__":
    asyncio.run(main())

File: elixir/examples/multimodal.exs
# Live multimodal example: attach an image, and let a TOOL return one.
#
# Run from the elixir/ directory (deps fetched):
#     mix run examples/multimodal.exs                          # offline: parts + blocks only
#     OPENROUTER_API_KEY=... mix run examples/multimodal.exs    # the live runs
#
# The key is read from the environment and never printed.
#
# TWO things are proved here, and only one of them is provable from the model's words:
#
#   1. THE IMAGE ARRIVED. The proof is the PROMPT-TOKEN DELTA between the identical
#      request without and with the image — not the answer. A model asked to name the
#      colours in an image it never received will name four colours anyway, confidently
#      and wrongly, which is exactly how the emission bug this release fixes stayed
#      hidden. Tokens come from the provider; the model's prose does not.
#   2. THE §8A RELOCATION RULE. A tool returns an image in its result parts; the loop
#      must relocate those parts onto the following user turn (a tool_result message
#      cannot carry an image block on either style) and the run must complete.

here = Path.dirname(__ENV__.file)
examples = Path.expand(Path.join([here, "..", "..", "examples"]))
fixture = Path.join([examples, "media", "fixture.png"])

alias Toolnexus.{Client, ContentPart, ToolResult}

question =
  "This is an 8x8 image with four solid quadrants. Name the colour of each quadrant " <>
    "in the order top-left, top-right, bottom-left, bottom-right. Answer with four " <>
    "colour words only."

colours = ~w(red green blue white)

# How many of the four fixture colours a piece of prose actually names.
named = fn text ->
  down = String.downcase(text || "")
  Enum.count(colours, &String.contains?(down, &1))
end

image = ContentPart.image!(fixture)
IO.puts("fixture: #{inspect(image)}")
IO.puts("openai block:    #{inspect(ContentPart.to_block(image, "openai") |> elem(1) |> Map.keys())}")
IO.puts("anthropic block: #{inspect(ContentPart.to_block(image, "anthropic") |> elem(1) |> Map.keys())}")

# A tool whose result carries the image — the §8A relocation path (SPEC §1B/§8A).
show =
  Toolnexus.define_tool(
    name: "show_fixture",
    description: "Return the 8x8 four-quadrant fixture image so you can look at it.",
    input_schema: %{"type" => "object", "properties" => %{}},
    execute: fn _args, _ctx ->
      %ToolResult{
        output: "the fixture image (image/png, 82 bytes)",
        parts: [ContentPart.image!(fixture)]
      }
    end
  )

tk = Toolnexus.create_toolkit!(builtins: false)
tk = Toolnexus.Toolkit.register(tk, [show])
empty = Toolnexus.create_toolkit!(builtins: false)

styles = [
  {"openai", System.get_env("OPENROUTER_MODEL_OPENAI") || "openai/gpt-4o-mini"},
  {"anthropic", System.get_env("OPENROUTER_MODEL_ANTHROPIC") || "anthropic/claude-haiku-4.5"}
]

case System.get_env("OPENROUTER_API_KEY") do
  nil ->
    IO.puts("\n(no OPENROUTER_API_KEY set — skipping the live runs)")

  key ->
    for {style, model} <- styles do
      client =
        Client.create(
          base_url: "https://openrouter.ai/api/v1",
          style: style,
          model: model,
          api_key: key,
          max_turns: 4,
          request_params: %{"max_tokens" => 40}
        )

      # 1. the identical request, without and with the image.
      text_only = Client.run(client, [ContentPart.text(question)], empty)
      with_image = Client.run(client, [ContentPart.text(question), ContentPart.image!(fixture)], empty)

      ptok_text = text_only.usage.prompt_tokens
      ptok_image = with_image.usage.prompt_tokens
      delta = ptok_image - ptok_text

      # 2. the relocation path: the model calls the tool, the image comes back in
      #    its result parts, and the loop must relocate it and finish.
      relocated =
        Client.run(
          client,
          "Call show_fixture, then " <> question,
          tk
        )

      relocation =
        if relocated.status == "done" and relocated.tool_call_count > 0, do: "ok", else: "FAILED"

      IO.puts("")
      IO.puts("[#{style}] #{model}")
      IO.puts("  text only  (#{ptok_text} ptok): #{String.replace(text_only.text, "\n", " ")}")
      IO.puts("  with image (#{ptok_image} ptok): #{String.replace(with_image.text, "\n", " ")}")
      IO.puts("  via tool   (#{relocated.tool_call_count} tool calls): #{String.replace(relocated.text, "\n", " ")}")

      IO.puts(
        "RESULT elixir style=#{style} ptok_text=#{ptok_text} ptok_image=#{ptok_image} " <>
          "delta=#{if delta >= 0, do: "+", else: ""}#{delta} " <>
          "colours=#{named.(with_image.text)}/4 relocation=#{relocation} " <>
          "reloc_colours=#{named.(relocated.text)}/4"
      )
    end
end

Toolnexus.Toolkit.close(tk)
Toolnexus.Toolkit.close(empty)
IO.puts("\nOK")
